### DISCUSSION TITLE
How to get ifc structural tree from parsing ifc file?

### DISCUSSION THREAD

**Discussion by sennhvi** [IfcOpenShell/IfcOpenShell#4141] [General]
I'd like to parse a ifc file using ifcopenshell python library. 

As far as I know, ifc uses relationships like IfcRelAggregates and IFCRELCONTAINEDINSPATIALSTRUCTURE to construct a hierarchy tree, what I want is for any entity_instance I get, I can get its location information like which IFCPROJECT/IFCSITE/IFCBUILDING/IFCBUILDINGSTOREY it locates. 

And it seems that relationships like IFCRELVOIDSELEMENT and IFCRELFILLSELEMENT also should be taken into account.

**Comment by aothms** [comment]
For the most part you could use this: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/util/element.py#L675

But it doesn't seem to include IFCRELVOIDSELEMENT IFCRELFILLSELEMENT  (accessed by the inverses IfcFeatureElementSubtraction,VoidsElements IfcElement.FillsVoids)

I think it would make a great extension to the get_container() function to include a `should_include_voids` or something and then prioritize the relationships above over the `ContainedInStructure` inverse. Pull Requests very welcome :)

**Reply by sennhvi** [reply]
Thanks for your reply! Recently I tried to use these 5 IFC relationships to describe the positioning of an entity_instance: IFCRELAGGREGATES \ IFCRELCONTAINEDINSPATIALSTRUCTURE \ IFCRELFILLSELEMENT \  IFCRELVOIDSELEMENT  \ IFCRELSPACEBOUNDARY.

And noticed that there is duplication of entities associated with these relationships, which makes it harder to decide which storey a window located in .

For example,  BUILDING is `#115` has `#132` and `#136` as BUILDINGSTOREY, in IFCRELCONTAINEDINSPATIALSTRUCTURE `#132` has a member `#234824 ` as window, `#136` has `#61565` as IFCWALLSTANDARDCASE, `#61565` contains a OPENINGELEMENT `#919798 `, `#919798 ` has `#234824 ` to fill.
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/31751605/46f80119-d59c-4133-8422-38583c252621)

This confuses me about  which storey the window #234824 belong to.

**Reply by aothms** [reply]
This is simply true, combining the containment and void+fill information results into two incoming edges for the window, which may in rare cases be different. There are  no real answers in terms of how to resolve that.
